homeassistant/components/zha/__init__.py

```diff
@@ -158,42 +161,67 @@ async def async_setup_entry(hass: HomeAssistant, config_entry: ConfigEntry) -> b
 
     _LOGGER.debug("Trigger cache: %s", zha_data.device_trigger_cache)
 
-    zha_gateway = ZHAGateway(hass, zha_data.yaml_config, config_entry)
+    # Retry setup a few times before giving up to deal with missing serial ports in VMs
+    for attempt in range(STARTUP_RETRIES):
+        try:
+            zha_gateway = await ZHAGateway.async_from_config(
+                hass=hass,
+                config=zha_data.yaml_config,
+                config_entry=config_entry,
+            )
+            break
+        except NetworkSettingsInconsistent as exc:
+            await warn_on_inconsistent_network_settings(
+                hass,
+                config_entry=config_entry,
+                old_state=exc.old_state,
+                new_state=exc.new_state,
+            )
+            raise HomeAssistantError(
```

Comment:
If we don't want to try again the correct exception to raise during config entry setup is `ConfigEntryError` (for known exceptions).